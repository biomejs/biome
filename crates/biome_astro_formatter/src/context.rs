use biome_formatter::prelude::*;
use biome_formatter::{
    comments::Comments, CstFormatContext, FormatContext, FormatOptions, GroupId,
    TransformSourceMap,
};
use biome_astro_syntax::AstroLanguage;
use std::rc::Rc;

#[derive(Debug, Clone)]
pub struct AstroFormatOptions {
    /// The indent style.
    indent_style: IndentStyle,

    /// The indent width.
    indent_width: IndentWidth,

    /// The line width.
    line_width: LineWidth,

    /// The line ending.
    line_ending: LineEnding,

    /// Astro-specific options
    astro_allow_shorthand: bool,
}

impl Default for AstroFormatOptions {
    fn default() -> Self {
        Self {
            indent_style: IndentStyle::default(),
            indent_width: IndentWidth::default(),
            line_width: LineWidth::default(),
            line_ending: LineEnding::default(),
            astro_allow_shorthand: true,
        }
    }
}

impl AstroFormatOptions {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn with_indent_style(mut self, indent_style: IndentStyle) -> Self {
        self.indent_style = indent_style;
        self
    }

    pub fn with_indent_width(mut self, indent_width: IndentWidth) -> Self {
        self.indent_width = indent_width;
        self
    }

    pub fn with_line_width(mut self, line_width: LineWidth) -> Self {
        self.line_width = line_width;
        self
    }

    pub fn with_line_ending(mut self, line_ending: LineEnding) -> Self {
        self.line_ending = line_ending;
        self
    }

    pub fn with_astro_allow_shorthand(mut self, allow_shorthand: bool) -> Self {
        self.astro_allow_shorthand = allow_shorthand;
        self
    }

    pub fn astro_allow_shorthand(&self) -> bool {
        self.astro_allow_shorthand
    }
}

impl FormatOptions for AstroFormatOptions {
    fn indent_style(&self) -> IndentStyle {
        self.indent_style
    }

    fn indent_width(&self) -> IndentWidth {
        self.indent_width
    }

    fn line_width(&self) -> LineWidth {
        self.line_width
    }

    fn line_ending(&self) -> LineEnding {
        self.line_ending
    }

    fn attribute_position(&self) -> AttributePosition {
        AttributePosition::Auto
    }

    fn bracket_spacing(&self) -> BracketSpacing {
        BracketSpacing::default()
    }

    fn as_print_options(&self) -> PrinterOptions {
        PrinterOptions::from(self)
    }
}

impl From<&AstroFormatOptions> for PrinterOptions {
    fn from(options: &AstroFormatOptions) -> Self {
        PrinterOptions::default()
            .with_indent(options.indent_style())
            .with_indent_width(options.indent_width())
            .with_line_width(options.line_width())
            .with_line_ending(options.line_ending())
    }
}

pub struct AstroFormatContext {
    options: AstroFormatOptions,
    comments: Rc<Comments<AstroLanguage>>,
    source_map: Option<TransformSourceMap>,
}

impl AstroFormatContext {
    pub fn new(options: AstroFormatOptions, comments: Comments<AstroLanguage>) -> Self {
        Self {
            options,
            comments: Rc::new(comments),
            source_map: None,
        }
    }

    pub fn with_source_map(mut self, source_map: Option<TransformSourceMap>) -> Self {
        self.source_map = source_map;
        self
    }
}

impl FormatContext for AstroFormatContext {
    type Options = AstroFormatOptions;

    fn options(&self) -> &Self::Options {
        &self.options
    }

    fn source_map(&self) -> Option<&TransformSourceMap> {
        self.source_map.as_ref()
    }
}

impl CstFormatContext for AstroFormatContext {
    type Language = AstroLanguage;

    fn comments(&self) -> &Comments<Self::Language> {
        &self.comments
    }
}