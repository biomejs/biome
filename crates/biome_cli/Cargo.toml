[package]
authors.workspace    = true
edition.workspace    = true
license.workspace    = true
name                 = "biome_cli"
repository.workspace = true
version              = "0.0.0"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
[[bin]]
name = "biome"
path = "src/main.rs"

[dependencies]
anyhow               = "1.0.52"
biome_console        = { workspace = true }
biome_deserialize    = { workspace = true }
biome_diagnostics    = { workspace = true }
biome_flags          = { workspace = true }
biome_formatter      = { workspace = true }
biome_fs             = { workspace = true }
biome_json_formatter = { workspace = true }
biome_json_parser    = { workspace = true }
biome_json_syntax    = { workspace = true }
biome_lsp            = { workspace = true }
biome_migrate        = { workspace = true }
biome_rowan          = { workspace = true }
biome_service        = { workspace = true }
biome_text_edit      = { workspace = true }
biome_text_size      = { workspace = true }
bpaf                 = { workspace = true }
crossbeam            = "0.8.1"
dashmap              = { workspace = true }
hdrhistogram         = { version = "7.5.0", default-features = false }
indexmap             = { workspace = true }
lazy_static          = { workspace = true }
rayon                = "1.5.1"
serde                = { workspace = true, features = ["derive"] }
serde_json           = { workspace = true }
tokio                = { workspace = true, features = ["io-std", "io-util", "net", "time", "rt", "sync", "rt-multi-thread", "macros"] }
tracing              = { workspace = true }
tracing-appender     = "0.2"
tracing-subscriber   = { version = "0.3.16", features = ["env-filter"] }
tracing-tree         = "0.2.2"

[target.'cfg(unix)'.dependencies]
libc  = "0.2.127"
tokio = { workspace = true, features = ["process"] }

[target.'cfg(windows)'.dependencies]
mimalloc = "0.1.29"

[target.'cfg(not(windows))'.dependencies]
tikv-jemallocator = "0.5.0"

[dev-dependencies]
biome_js_formatter   = { workspace = true }
biome_json_formatter = { workspace = true }
biome_json_parser    = { workspace = true }
insta                = { workspace = true }
tokio                = { workspace = true, features = ["io-util"] }

[features]
docgen = ["bpaf/docgen"]
