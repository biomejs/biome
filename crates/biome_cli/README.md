# `biome_cli`

The main binary distribution, exposes the command line interface defined in this crate,
and the language server interface defined in `biome_lsp` and used by the `biome` VSCode extension

# Logs

When the server is run in daemon mode,
it will output logs to a file created in the `biome-logs` directory inside the system temporary directory.
The log file will be rotated on a hourly basis.
