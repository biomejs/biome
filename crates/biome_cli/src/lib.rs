#![doc = include_str!("../README.md")]
//!
//! # Module
//!
//! This is where the main CLI session starts. The module is responsible
//! to parse commands and arguments, redirect the execution of the commands and
//! execute the traversal of directory and files, based on the command that were passed.

use biome_console::{ColorMode, Console};
use biome_fs::OsFileSystem;
use biome_service::{App, DynRef, Workspace, WorkspaceRef};
use std::env;

mod cli_options;
mod commands;
mod configuration;
mod diagnostics;
mod execute;
mod metrics;
mod panic;
mod reports;
mod service;
mod vcs;

use crate::cli_options::ColorsArg;
use crate::commands::check::CheckCommandPayload;
use crate::commands::ci::CiCommandPayload;
use crate::commands::format::FormatCommandPayload;
use crate::commands::lint::LintCommandPayload;
pub use crate::commands::{biome_command, BiomeCommand};
pub use diagnostics::CliDiagnostic;
pub(crate) use execute::{execute_mode, Execution, TraversalMode};
pub use panic::setup_panic_handler;
pub use reports::{
    formatter::{FormatterReport, FormatterReportFileDetail, FormatterReportSummary},
    Report, ReportDiagnostic, ReportDiff, ReportErrorKind, ReportKind,
};
pub use service::{open_transport, SocketTransport};

pub(crate) const VERSION: &str = match option_env!("BIOME_VERSION") {
    Some(version) => version,
    None => env!("CARGO_PKG_VERSION"),
};

/// Global context for an execution of the CLI
pub struct CliSession<'app> {
    /// Instance of [App] used by this run of the CLI
    pub app: App<'app>,
}

impl<'app> CliSession<'app> {
    pub fn new(
        workspace: &'app dyn Workspace,
        console: &'app mut dyn Console,
    ) -> Result<Self, CliDiagnostic> {
        Ok(Self {
            app: App::new(
                DynRef::Owned(Box::new(OsFileSystem)),
                console,
                WorkspaceRef::Borrowed(workspace),
            ),
        })
    }

    /// Main function to run Biome CLI
    pub fn run(self, command: BiomeCommand) -> Result<(), CliDiagnostic> {
        let has_metrics = command.has_metrics();
        if has_metrics {
            crate::metrics::init_metrics();
        }

        let result = match command {
            BiomeCommand::Version(_) => commands::version::full_version(self),
            BiomeCommand::Rage(_) => commands::rage::rage(self),
            BiomeCommand::Start => commands::daemon::start(self),
            BiomeCommand::Stop => commands::daemon::stop(self),
            BiomeCommand::Check {
                apply,
                apply_unsafe,
                cli_options,
                configuration: rome_configuration,
                paths,
                stdin_file_path,
                linter_enabled,
                organize_imports_enabled,
                formatter_enabled,
            } => commands::check::check(
                self,
                CheckCommandPayload {
                    apply_unsafe,
                    apply,
                    cli_options,
                    configuration: rome_configuration,
                    paths,
                    stdin_file_path,
                    linter_enabled,
                    organize_imports_enabled,
                    formatter_enabled,
                },
            ),
            BiomeCommand::Lint {
                apply,
                apply_unsafe,
                cli_options,
                linter_configuration,
                paths,
                stdin_file_path,
                vcs_configuration,
                files_configuration,
            } => commands::lint::lint(
                self,
                LintCommandPayload {
                    apply_unsafe,
                    apply,
                    cli_options,
                    linter_configuration,
                    paths,
                    stdin_file_path,
                    vcs_configuration,
                    files_configuration,
                },
            ),
            BiomeCommand::Ci {
                linter_enabled,
                formatter_enabled,
                organize_imports_enabled,
                configuration: rome_configuration,
                paths,
                cli_options,
            } => commands::ci::ci(
                self,
                CiCommandPayload {
                    linter_enabled,
                    formatter_enabled,
                    organize_imports_enabled,
                    rome_configuration,
                    paths,
                    cli_options,
                },
            ),
            BiomeCommand::Format {
                javascript_formatter,
                formatter_configuration,
                stdin_file_path,
                write,
                cli_options,
                paths,
                vcs_configuration,
                files_configuration,
                json_formatter,
            } => commands::format::format(
                self,
                FormatCommandPayload {
                    javascript_formatter,
                    formatter_configuration,
                    stdin_file_path,
                    write,
                    cli_options,
                    paths,
                    vcs_configuration,
                    files_configuration,
                    json_formatter,
                },
            ),
            BiomeCommand::Init => commands::init::init(self),
            BiomeCommand::LspProxy(_) => commands::daemon::lsp_proxy(),
            BiomeCommand::Migrate(cli_options, write) => {
                commands::migrate::migrate(self, cli_options, write)
            }
            BiomeCommand::RunServer { stop_on_disconnect } => {
                commands::daemon::run_server(stop_on_disconnect)
            }
            BiomeCommand::PrintSocket => commands::daemon::print_socket(),
        };

        if has_metrics {
            metrics::print_metrics();
        }

        result
    }
}

pub fn to_color_mode(color: Option<&ColorsArg>) -> ColorMode {
    match color {
        Some(ColorsArg::Off) => ColorMode::Disabled,
        Some(ColorsArg::Force) => ColorMode::Enabled,
        None => ColorMode::Auto,
    }
}
