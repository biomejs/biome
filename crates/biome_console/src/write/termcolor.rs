use std::{
    fmt::{self, Write as _},
    io,
};

use termcolor::{Color, ColorSpec, WriteColor};
use unicode_segmentation::UnicodeSegmentation;
use unicode_width::UnicodeWidthStr;

use crate::{fmt::MarkupElements, MarkupElement};

use super::Write;

/// Adapter struct implementing [Write] over types implementing [WriteColor]
pub struct Termcolor<W>(pub W);

impl<W> Write for Termcolor<W>
where
    W: WriteColor,
{
    fn write_str(&mut self, elements: &MarkupElements, content: &str) -> io::Result<()> {
        with_format(&mut self.0, elements, |writer| {
            let mut adapter = SanitizeAdapter {
                writer,
                error: Ok(()),
            };

            match adapter.write_str(content) {
                Ok(()) => Ok(()),
                Err(..) => {
                    if adapter.error.is_err() {
                        adapter.error
                    } else {
                        // SanitizeAdapter can only fail if the underlying
                        // writer returns an error
                        unreachable!()
                    }
                }
            }
        })
    }

    fn write_fmt(&mut self, elements: &MarkupElements, content: fmt::Arguments) -> io::Result<()> {
        with_format(&mut self.0, elements, |writer| {
            let mut adapter = SanitizeAdapter {
                writer,
                error: Ok(()),
            };

            match adapter.write_fmt(content) {
                Ok(()) => Ok(()),
                Err(..) => {
                    if adapter.error.is_err() {
                        adapter.error
                    } else {
                        Err(io::Error::new(
                            io::ErrorKind::Other,
                            "a Display formatter returned an error",
                        ))
                    }
                }
            }
        })
    }
}

/// Applies the current format in `state` to `writer`, calls `func` to
/// print a piece of text, then reset the printing format
fn with_format<W>(
    writer: &mut W,
    state: &MarkupElements,
    func: impl FnOnce(&mut W) -> io::Result<()>,
) -> io::Result<()>
where
    W: WriteColor,
{
    let mut color = ColorSpec::new();
    let mut link = None;
    let mut inverse = false;

    state.for_each(&mut |elements| {
        for element in elements {
            match element {
                MarkupElement::Inverse => {
                    inverse = !inverse;
                }
                MarkupElement::Hyperlink { href } => {
                    link = Some(href);
                }
                _ => {
                    element.update_color(&mut color);
                }
            }
        }

        Ok(())
    })?;

    if inverse {
        let fg = color.fg().map_or(Color::White, |c| *c);
        let bg = color.bg().map_or(Color::Black, |c| *c);
        color.set_bg(Some(fg));
        color.set_fg(Some(bg));
    }

    if let Err(err) = writer.set_color(&color) {
        writer.reset()?;
        return Err(err);
    }

    let mut reset_link = false;
    if let Some(href) = link {
        // `is_synchronous` is used to check if the underlying writer
        // is using the Windows Console API, that does not support ANSI
        // escape codes. Generally this would only be true when running
        // in the legacy `cmd.exe` terminal emulator, since in modern
        // clients like the Windows Terminal ANSI is used instead
        if writer.supports_color() && !writer.is_synchronous() {
            write!(writer, "\x1b]8;;{href}\x1b\\")?;
            reset_link = true;
        }
    }

    let result = func(writer);

    if reset_link {
        write!(writer, "\x1b]8;;\x1b\\")?;
    }

    writer.reset()?;
    result
}

/// Adapter [fmt::Write] calls to [io::Write] with sanitization,
/// implemented as an internal struct to avoid exposing [fmt::Write] on
/// [Termcolor]
struct SanitizeAdapter<W> {
    writer: W,
    error: io::Result<()>,
}

impl<W> fmt::Write for SanitizeAdapter<W>
where
    W: WriteColor,
{
    fn write_str(&mut self, content: &str) -> fmt::Result {
        let mut buffer = [0; 4];

        for grapheme in content.graphemes(true) {
            let width = UnicodeWidthStr::width(grapheme);
            let is_whitespace = grapheme_is_whitespace(grapheme);

            if !is_whitespace && width == 0 {
                let char_to_write = char::REPLACEMENT_CHARACTER;
                char_to_write.encode_utf8(&mut buffer);

                if let Err(err) = self.writer.write_all(&buffer[..char_to_write.len_utf8()]) {
                    self.error = Err(err);
                    return Err(fmt::Error);
                }

                continue;
            }

            // Unicode is currently poorly supported on most Windows
            // terminal clients, so we always strip emojis in Windows
            if cfg!(windows) || !self.writer.supports_color() {
                let is_ascii = grapheme_is_ascii(grapheme);

                if !is_ascii {
                    let replacement = unicode_to_ascii(grapheme.chars().nth(0).unwrap());

                    replacement.encode_utf8(&mut buffer);

                    if let Err(err) = self.writer.write_all(&buffer[..replacement.len_utf8()]) {
                        self.error = Err(err);
                        return Err(fmt::Error);
                    }

                    continue;
                }
            };

            for char in grapheme.chars() {
                char.encode_utf8(&mut buffer);

                if let Err(err) = self.writer.write_all(&buffer[..char.len_utf8()]) {
                    self.error = Err(err);
                    return Err(fmt::Error);
                }
            }
        }

        Ok(())
    }
}

/// Determines if a unicode grapheme consists only of code points
/// which are considered whitespace characters in ASCII
fn grapheme_is_whitespace(grapheme: &str) -> bool {
    grapheme.chars().all(|c| c.is_whitespace())
}

/// Determines if a grapheme contains code points which are out of the ASCII
/// range and thus cannot be printed where unicode is not supported.
fn grapheme_is_ascii(grapheme: &str) -> bool {
    grapheme.chars().all(|c| c.is_ascii())
}

/// Replace emoji characters with similar but more widely supported ASCII
/// characters
fn unicode_to_ascii(c: char) -> char {
    match c {
        '\u{2714}' => '\u{221a}',
        '\u{2139}' => 'i',
        '\u{26a0}' => '!',
        '\u{2716}' => '\u{00d7}',
        _ => c,
    }
}

#[cfg(test)]
mod tests {
    use std::{fmt::Write, str::from_utf8};

    use biome_markup::markup;
    use termcolor::Ansi;

    use crate as biome_console;
    use crate::fmt::Formatter;

    use super::{SanitizeAdapter, Termcolor};

    #[test]
    fn test_sanitize() {
        // Sanitization should leave whitespace control characters (space,
        // tabs, newline, ...) and non-ASCII unicode characters as-is but
        // redact zero-width characters (RTL override, null character, bell,
        // zero-width space, ...)
        const INPUT: &str = "t\tes t\r\n\u{202D}t\0es\x07t\u{202E}\nt\u{200B}esüêõt";
        const OUTPUT: &str = "t\tes t\r\n\u{FFFD}t\u{FFFD}es\u{FFFD}t\u{FFFD}\nt\u{FFFD}esüêõt";

        let mut buffer = Vec::new();

        {
            let writer = termcolor::Ansi::new(&mut buffer);
            let mut adapter = SanitizeAdapter {
                writer,
                error: Ok(()),
            };

            adapter.write_str(INPUT).unwrap();
            adapter.error.unwrap();
        }

        assert_eq!(from_utf8(&buffer).unwrap(), OUTPUT);
    }

    #[test]
    fn test_hyperlink() {
        const OUTPUT: &str = "\x1b[0m\x1b]8;;https://biomejs.dev/\x1b\\link\x1b]8;;\x1b\\\x1b[0m";

        let mut buffer = Vec::new();
        let mut writer = Termcolor(Ansi::new(&mut buffer));
        let mut formatter = Formatter::new(&mut writer);

        formatter
            .write_markup(markup! {
                <Hyperlink href="https://biomejs.dev/">"link"</Hyperlink>
            })
            .unwrap();

        assert_eq!(from_utf8(&buffer).unwrap(), OUTPUT);
    }

    #[test]
    fn test_printing_complex_emojis() {
        const INPUT: &str = "‚ö†Ô∏è1Ô∏è‚É£‚ÑπÔ∏è";
        const OUTPUT: &str = "‚ö†Ô∏è1Ô∏è‚É£‚ÑπÔ∏è";
        const WINDOWS_OUTPUT: &str = "!1i";

        let mut buffer = Vec::new();

        {
            let writer = termcolor::Ansi::new(&mut buffer);
            let mut adapter = SanitizeAdapter {
                writer,
                error: Ok(()),
            };

            adapter.write_str(INPUT).unwrap();
            adapter.error.unwrap();
        }

        if cfg!(windows) {
            assert_eq!(from_utf8(&buffer).unwrap(), WINDOWS_OUTPUT);
        } else {
            assert_eq!(from_utf8(&buffer).unwrap(), OUTPUT);
        }
    }
}
