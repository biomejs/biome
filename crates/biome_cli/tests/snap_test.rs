use biome_cli::CliDiagnostic;
use biome_console::fmt::{Formatter, Termcolor};
use biome_console::{markup, BufferConsole, Markup};
use biome_diagnostics::termcolor::NoColor;
use biome_diagnostics::{print_diagnostic_to_string, Error};
use biome_formatter::{IndentStyle, IndentWidth};
use biome_fs::{ConfigName, FileSystemExt, MemoryFileSystem};
use biome_json_formatter::context::JsonFormatOptions;
use biome_json_formatter::format_node;
use biome_json_parser::{parse_json, JsonParserOptions};
use lazy_static::lazy_static;
use regex::Regex;
use std::borrow::Cow;
use std::collections::BTreeMap;
use std::env::{current_exe, temp_dir};
use std::fmt::Write as _;
use std::path::{Path, PathBuf, MAIN_SEPARATOR};

lazy_static! {
    static ref TIME_REGEX: Regex = Regex::new("\\s[0-9]+[mÂµn]?s\\.").unwrap();
    static ref TIME_JUNIT_REGEX: Regex = Regex::new("time=\\\"[.0-9]+\\\"").unwrap();
}

#[derive(Default)]
struct InMessages {
    stdin: Option<String>,
}

pub(crate) struct CliSnapshot {
    /// input messages, coming from different sources
    in_messages: InMessages,
    /// the configuration, if set
    /// First string is the content
    /// Second string is the name
    pub configuration: Option<(String, &'static str)>,
    /// file name -> content
    pub files: BTreeMap<String, String>,
    /// messages written in console
    pub messages: Vec<String>,
    /// possible termination error of the CLI
    pub termination: Option<Error>,
}

impl CliSnapshot {
    pub fn from_result(result: Result<(), CliDiagnostic>) -> Self {
        Self {
            in_messages: InMessages::default(),
            configuration: None,
            files: BTreeMap::default(),
            messages: Vec::new(),
            termination: result.err().map(Error::from),
        }
    }
}

impl CliSnapshot {
    pub fn emit_content_snapshot(&self) -> String {
        let mut content = String::new();

        if let Some((configuration, file_name)) = &self.configuration {
            let redacted = redact_snapshot(configuration).unwrap_or(String::new().into());

            let parsed = parse_json(
                &redacted,
                JsonParserOptions::default()
                    .with_allow_comments()
                    .with_allow_trailing_commas(),
            );
            let formatted = format_node(
                JsonFormatOptions::default()
                    .with_indent_style(IndentStyle::Space)
                    .with_indent_width(IndentWidth::default()),
                &parsed.syntax(),
            )
            .expect("formatted JSON")
            .print()
            .expect("printed JSON");

            content.push_str(&format!("## `{file_name}`\n\n"));
            content.push_str("```json");
            content.push('\n');
            content.push_str(formatted.as_code());
            content.push_str("```");
            content.push_str("\n\n")
        }

        for (name, file_content) in &self.files {
            if !name.starts_with("biome.json") {
                let extension = name.split('.').last().unwrap();

                let redacted_name = redact_snapshot(name).unwrap_or(String::new().into());
                let redacted_content =
                    redact_snapshot(file_content).unwrap_or(String::new().into());

                let _ = write!(content, "## `{}`\n\n", redacted_name);
                let _ = write!(content, "```{extension}");
                content.push('\n');
                content.push_str(&redacted_content);
                content.push('\n');
                content.push_str("```");
                content.push_str("\n\n")
            }
        }

        if let Some(stdin) = &self.in_messages.stdin {
            content.push_str("# Input messages\n\n");
            content.push_str("```block");
            content.push('\n');
            content.push_str(stdin);
            content.push('\n');
            content.push_str("```");
            content.push_str("\n\n")
        }

        if let Some(termination) = &self.termination {
            let message = print_diagnostic_to_string(termination);

            if let Some(redacted) = &redact_snapshot(&message) {
                content.push_str("# Termination Message\n\n");
                content.push_str("```block");
                content.push('\n');
                content.push_str(redacted);
                content.push('\n');
                content.push_str("```");
                content.push_str("\n\n");
            }
        }

        if !self.messages.is_empty() {
            content.push_str("# Emitted Messages\n\n");

            for message in &self.messages {
                let Some(redacted) = &redact_snapshot(message) else {
                    continue;
                };

                content.push_str("```block");
                content.push('\n');
                content.push_str(redacted);
                content.push('\n');
                content.push_str("```");
                content.push_str("\n\n")
            }
        }

        content
    }
}

fn redact_snapshot(input: &str) -> Option<Cow<'_, str>> {
    let mut output = Cow::Borrowed(input);

    // There are some logs that print the timing, and we can't snapshot that message
    // otherwise at each run we invalid the previous snapshot.
    //
    // This is a workaround, and it might not work for all cases.
    let the_match = TIME_REGEX.find(output.as_ref()).map(|f| f.start()..f.end());
    if let Some(found) = the_match {
        output.to_mut().replace_range(found, " <TIME>.");
    }

    let the_match = TIME_JUNIT_REGEX
        .find(output.as_ref())
        .map(|f| f.start()..f.end());
    if let Some(found) = the_match {
        output.to_mut().replace_range(found, "time=\"<TIME>\"");
    }

    // Normalize the name of the current executable to "biome"
    let current_exe = current_exe()
        .ok()
        .and_then(|path| Some(path.file_name()?.to_str()?.to_string()));

    if let Some(current_exe) = current_exe {
        replace(&mut output, &current_exe, "biome");
    }

    output = replace_temp_dir(output);
    output = replace_biome_dir(output);

    // Normalize Windows-specific path separators to "/"
    if cfg!(windows) {
        let mut rest = &*output;
        let mut result = String::new();

        while let Some(index) = rest.find(MAIN_SEPARATOR) {
            let (before, after) = rest.split_at(index);
            result.push_str(before);

            // Paths are recognized if they start with ".\",  ":\" (as in "C:\")
            // or ">\" (as in "<TEMP_DIR>\")
            if !before.ends_with(['.', ':', '>']) {
                let (sep, after) = after.split_at(1);
                result.push_str(sep);
                rest = after;
                continue;
            }

            // File paths are assumed to end at the first space or line breaks
            let path = if let Some(end) = after.find([' ', '\n']) {
                let (before, after) = after.split_at(end);
                rest = after;
                before
            } else {
                rest = "";
                after
            };

            result.push_str(&path.replace(MAIN_SEPARATOR, "/"));
        }

        if !result.is_empty() {
            result.push_str(&rest.replace(MAIN_SEPARATOR, "/"));
            output = Cow::Owned(result);
        }
    }

    Some(output)
}

/// Replace the path to the temporary directory with "<TEMP_DIR>"
/// And normalizes the count of `-` at the end of the diagnostic
fn replace_temp_dir(input: Cow<str>) -> Cow<str> {
    let mut result = String::new();
    let mut rest = input.as_ref();

    let temp_dir = temp_dir().display().to_string();
    let temp_dir = temp_dir.trim_end_matches(MAIN_SEPARATOR);

    while let Some(index) = rest.find(temp_dir) {
        let (before, after) = rest.split_at(index);

        result.push_str(before);
        result.push_str("<TEMP_DIR>");

        let after = after.split_at(temp_dir.len()).1;
        let header_line = after.lines().next().unwrap();

        match header_line.split_once('\u{2501}') {
            Some((between_temp_and_line, _)) => {
                // Diagnostic header line, normalize the horizontal line
                result.push_str(between_temp_and_line);
                result.push_str(&"\u{2501}".repeat(20));
                rest = after.split_at(header_line.len()).1;
            }
            None => {
                // Not a header line, only replace tempdir
                rest = after;
            }
        }
    }

    if result.is_empty() {
        input
    } else {
        result.push_str(rest);
        Cow::Owned(result)
    }
}

/// Replace the path to the temporary directory with "<TEMP_DIR>"
/// And normalizes the count of `-` at the end of the diagnostic
fn replace_biome_dir(input: Cow<str>) -> Cow<str> {
    let mut result = String::new();
    let mut rest = input.as_ref();

    let temp_dir = biome_fs::ensure_cache_dir().display().to_string();
    let temp_dir = temp_dir.trim_end_matches(MAIN_SEPARATOR);

    while let Some(index) = rest.find(temp_dir) {
        let (before, after) = rest.split_at(index);

        result.push_str(before);
        result.push_str("<BIOME_DIR>");

        let after = after.split_at(temp_dir.len()).1;
        let header_line = after.lines().next().unwrap();

        match header_line.split_once('\u{2501}') {
            Some((between_temp_and_line, _)) => {
                // Diagnostic header line, normalize the horizontal line
                result.push_str(between_temp_and_line);
                result.push_str(&"\u{2501}".repeat(20));
                rest = after.split_at(header_line.len()).1;
            }
            None => {
                // Not a header line, only replace tempdir
                rest = after;
            }
        }
    }

    if result.is_empty() {
        input
    } else {
        result.push_str(rest);
        Cow::Owned(result)
    }
}

fn replace(input: &mut Cow<str>, from: &str, to: &str) {
    let mut rest = &**input;
    let mut result = String::new();

    while let Some(index) = rest.find(from) {
        let (before, after) = rest.split_at(index);

        result.push_str(before);
        result.push_str(to);

        let (_, after) = after.split_at(from.len());
        rest = after;
    }

    if !result.is_empty() {
        result.push_str(rest);
        *input = Cow::Owned(result);
    }
}

impl From<SnapshotPayload<'_>> for CliSnapshot {
    fn from(payload: SnapshotPayload<'_>) -> Self {
        let SnapshotPayload {
            result,
            console,
            fs,
            test_name: _,
            module_path: _,
        } = payload;
        let mut cli_snapshot = CliSnapshot::from_result(result);

        for file_name in ConfigName::file_names() {
            let config_path = PathBuf::from(file_name);
            let configuration = fs.open(&config_path).ok();
            if let Some(mut configuration) = configuration {
                let mut buffer = String::new();
                if configuration.read_to_string(&mut buffer).is_ok() {
                    cli_snapshot.configuration = Some((buffer, file_name));
                }
            }
        }

        let files: Vec<_> = fs
            .files()
            .map(|(file, entry)| {
                let content = entry.lock();
                let content = std::str::from_utf8(content.as_slice()).unwrap();
                (file.to_str().unwrap().to_string(), String::from(content))
            })
            .collect();

        for (file, content) in files {
            cli_snapshot.files.insert(file, content);
        }

        let in_buffer = &console.in_buffer;
        for (index, message) in in_buffer.iter().enumerate() {
            if index == 0 {
                cli_snapshot.in_messages.stdin = Some(message.to_string());
            }
        }

        for message in &console.out_buffer {
            let content = markup_to_string(markup! {
                {message.content}
            });
            cli_snapshot.messages.push(content)
        }

        cli_snapshot
    }
}

pub fn markup_to_string(markup: Markup) -> String {
    let mut buffer = Vec::new();
    let mut write = Termcolor(NoColor::new(&mut buffer));
    let mut fmt = Formatter::new(&mut write);
    fmt.write_markup(markup).unwrap();

    String::from_utf8(buffer).unwrap()
}

pub struct SnapshotPayload<'a> {
    pub module_path: &'a str,
    pub test_name: &'a str,
    pub fs: MemoryFileSystem,
    pub console: BufferConsole,
    pub result: Result<(), CliDiagnostic>,
}

impl<'a> SnapshotPayload<'a> {
    pub fn new(
        module_path: &'a str,
        test_name: &'a str,
        fs: MemoryFileSystem,
        console: BufferConsole,
        result: Result<(), CliDiagnostic>,
    ) -> Self {
        Self {
            module_path,
            test_name,
            fs,
            console,
            result,
        }
    }
}

/// Function used to snapshot a session test of the a CLI run.
pub fn assert_cli_snapshot(payload: SnapshotPayload<'_>) {
    let module_path = payload.module_path.to_owned();
    let test_name = payload.test_name;
    let cli_snapshot = CliSnapshot::from(payload);

    let content = cli_snapshot.emit_content_snapshot();

    let module_path = module_path.replace("::", "_");
    let snapshot_path = PathBuf::from("snapshots").join(module_path);

    insta::with_settings!({
        prepend_module_to_snapshot => false,
        snapshot_path => snapshot_path
    }, {
        insta::assert_snapshot!(test_name, content);

    });
}

/// It checks if the contents of a file matches the passed `expected_content`
pub fn assert_file_contents(fs: &MemoryFileSystem, path: &Path, expected_content: &str) {
    let mut file = fs.open(path).expect("file was removed");

    let mut content = String::new();
    file.read_to_string(&mut content)
        .expect("failed to read file from memory FS");

    assert_eq!(
        content,
        expected_content,
        "file {} doesn't match the expected content (right)",
        path.display()
    );
}
