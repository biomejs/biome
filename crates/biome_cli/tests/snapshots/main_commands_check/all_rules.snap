---
source: crates/biome_cli/tests/snap_test.rs
expression: content
---
## `biome.json`

```json
{
  "linter": {
    "rules": { "all": true }
  }
}
```

## `fix.js`

```js

(1 >= -0)

```

# Termination Message

```block
check ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  × Some errors were emitted while running checks.
  


```

# Emitted Messages

```block
fix.js:2:2 lint/suspicious/noCompareNegZero  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  × Do not use the >= operator to compare against -0.
  
  > 2 │ (1 >= -0)
      │  ^^^^^^^
    3 │ 
  
  i Safe fix: Replace -0 with 0
  
    2 │ (1·>=·-0)
      │       -  

```

```block
fix.js lint ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  × The file contains diagnostics that needs to be addressed.
  

```

```block
fix.js format ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  i Formatter would have printed the following content:
  
    1   │ - 
    2   │ - (1·>=·-0)
      1 │ + 1·>=·-0;
    3 2 │   
  

```

```block
fix.js check ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  × The file contains diagnostics that needs to be addressed.
  

```

```block
Checked 1 file(s) in <TIME>
```


